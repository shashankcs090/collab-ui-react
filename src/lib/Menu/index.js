import React from 'react';
import PropTypes from 'prop-types';
import { clone, take, isEqual, set, dropRight, concat, last, tail } from 'lodash';

/**
 * Menu is container component which contains MenuItems
 * @category containers
 * @component Menu
 * @variations collab-ui-react
 */

export default class Menu extends React.Component {
  static displayName = 'Menu';

  state = {
    menuIndex: [],
  };

  static childContextTypes = {
    handleClick: PropTypes.func,
    handleKeyDown: PropTypes.func
  };

  getChildContext = () => {
    return {
      handleClick: (menuItemIndex, menuItem, e) => this.handleClick(menuItemIndex, menuItem, e),
      handleKeyDown: (menuItemIndex, menuItem, e) => this.handleKeyDown(menuItemIndex, menuItem, e),
    };
  }

  componentWillMount() {
    const { children } = this.props;
    this.setState({
      menuIndex: [this.getNewIndex(-1, 1, React.Children.toArray(children))],
    });
  }

  handleClick = (menuItemIndex, menuItem, e) => {
    const { onSelect } = this.props;
    if(onSelect) {
      onSelect(e, menuItemIndex, menuItem);
    }
    this.setState({
      menuIndex: menuItem.props.subMenu ?
        concat(menuItemIndex, this.getNewIndex(-1, 1, menuItem.props.subMenu)) : [menuItemIndex[0]],
    });
  };

  getNewIndex = (currentIndex, change, siblings) => {
    const getPossibleIndex = () => {
      const length = siblings.length - 1;
      if (currentIndex + change < 0) {
        return length;
      } else if (currentIndex + change > length) {
        return 0;
      }

      return currentIndex + change;
    };

    const possibleIndex = getPossibleIndex();
    const potentialTarget = React.Children.toArray(siblings)[possibleIndex];

    return (potentialTarget.props.disabled || potentialTarget.props.isHeader)
      ? this.getNewIndex(possibleIndex, change, siblings)
      : possibleIndex;
  };

  handleKeyDown = (idx, menuItem, e) => {

    const { menuIndex } = this.state;
    const { subMenu } = menuItem.props;
    const { children, onSelect } = this.props;

    const getSiblings = (idx, menuItems = children) => {
      if(idx.length === 1) {
        return menuItems;
      }
      return getSiblings(tail(idx), menuItems[idx[0]].props.subMenu);
    };

    let clickEvent;
    const tgt = e.currentTarget;
    let flag = false;
    switch (e.which) {
      case 32:
      case 13:
        try {
          clickEvent = new MouseEvent('click', {
            view: window,
            bubbles: true,
            cancelable: true,
          });
        } catch (err) {
          if (document.createEvent) {
            // DOM Level 3 for IE 9+
            clickEvent = document.createEvent('MouseEvents');
            clickEvent.initEvent('click', true, true);
          }
        }
        tgt.dispatchEvent(clickEvent);

        flag = true;
        break;

      case 38:
        this.setState({
          menuIndex:
            set(clone(menuIndex), menuIndex.length - 1, this.getNewIndex(last(idx), -1, getSiblings(idx))),
        });
        flag = true;
        break;

      case 40:
        this.setState({
          menuIndex:
            set(clone(menuIndex), menuIndex.length - 1, this.getNewIndex(last(idx), 1, getSiblings(idx))),
        });
        flag = true;
        break;

      case 39: //right
        subMenu && this.setState({
          menuIndex: concat(menuIndex, this.getNewIndex(-1, 1, subMenu)),
        }, () => onSelect && onSelect(e, menuIndex, menuItem));

        flag = true;
        break;

      case 37: //left
        (menuIndex.length - 1) && this.setState({
          menuIndex: dropRight(menuIndex),
        });
        flag = true;
        break;

      default:
        break;
    }

    if (flag) {
      e.stopPropagation();
      e.preventDefault();
    }
  };

  render() {
    const { children, className, ariaLabel } = this.props;
    const setMenuItems = (menuList, currentItemIndex = []) => {
      const menuItems = React.Children.toArray(menuList).map((child, idx) => {
        if(child.type.displayName !== 'MenuItem') {
          throw new Error('child elements and subMenu items should be MenuItem');
        }
        const menuItemIndex = currentItemIndex.concat(idx);
        const { subMenu, disabled } = child.props;
        const { menuIndex } = this.state;

        const focus = isEqual(menuIndex, menuItemIndex);
        const isOpen = isEqual(take(menuIndex, menuItemIndex.length), menuItemIndex) &&
            menuIndex.length !== menuItemIndex.length;

        return React.cloneElement(child, {
          subMenu: subMenu && setMenuItems(subMenu, menuItemIndex),
          index: menuItemIndex,
          focus,
          isOpen,
        });
      });
      return menuItems;
    };

    return (
      <div
        className={
          'cui-menu' +
          `${(className && ` ${className}`) || ''}`
        }
        aria-label={ariaLabel}
        role="menubar"
      >
        {setMenuItems(children)}
      </div>
    );
  }
}

Menu.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  onSelect: PropTypes.func,
  ariaLabel: PropTypes.string,
};

Menu.defaultProps = {
  children: null,
  className: '',
  onSelect: null,
  ariaLabel: '',
};

/**
* @name Menu with Popover
*
* @category containers
* @component menu
* @section default
*
* @js
*
import { MenuItem, Button, Popover } from '@collab-ui/react';

export default class MenuDefault extends React.PureComponent {
  render() {
    const submenu1 = [
      <MenuItem key="0" isHeader>Set Do Not Disturb:</MenuItem>,
      <MenuItem key="1" disabled>1 hour</MenuItem>,
      <MenuItem key="2">5 hour</MenuItem>,
      <MenuItem key="3">8 hour</MenuItem>
    ];
    const submenu2 = [<MenuItem key="0">English</MenuItem>, <MenuItem key="1">Spanish</MenuItem>];
    const popoverContent = (
      <Menu>
        <MenuItem
          key="0"
          selectedValue="Out of office until 2:00pm"
          subMenu={submenu1}
        >
          Status
        </MenuItem>
        <MenuItem
          key="1"
          subMenu={submenu2}
          selectedValue="English"
        >
          Language
        </MenuItem>
        <MenuItem key="2">
          Settings
        </MenuItem>
      </Menu>
    );
    return(
      <Popover content={popoverContent} popoverTrigger="Click" direction="bottom-center">
        <Button>Show Menu</Button>
      </Popover>
    );
  }
}
**/

/**
* @name Customized menu with EventOverlay
*
* @category containers
* @component menu
* @section custom-menu
*
* @js
*
 import { MenuItem, Button, EventOverlay } from '@collab-ui/react';

 export default class CustomMenu extends React.PureComponent {
  state = {
    isOpen: false
  };

  onSelect = (e, idx, menuItem) => {
    if(menuItem.props.shouldCloseMenu) {
      this.setState({ isOpen: false });
    }
  };

  render() {
    const { isOpen } = this.state;
    const submenu1 = [
      <MenuItem key="0" isHeader>Set Do Not Disturb:</MenuItem>,
      <MenuItem key="1" disabled>1 hour</MenuItem>,
      <MenuItem key="2">5 hour</MenuItem>,
      <MenuItem key="3">8 hour</MenuItem>
    ];
    const submenu2 = [<MenuItem key="0" shouldCloseMenu>English</MenuItem>, <MenuItem key="1" shouldCloseMenu>Spanish</MenuItem>];
    const popoverContent = (
      <Menu onSelect={this.onSelect}>
        <MenuItem key="0" subMenu={submenu1}>
          Status
        </MenuItem>
        <MenuItem key="1" subMenu={submenu2}>
          Language
        </MenuItem>
        <MenuItem key="2" shouldCloseMenu>
          Settings
        </MenuItem>
      </Menu>
    );
    return(
      <div>
        <Button
          onClick={() => this.setState({ isOpen: !this.state.isOpen})}
          ref={ref => this.anchor = ref}
        >
          Show Menu
        </Button>
        <EventOverlay
          allowClickAway
          isOpen={isOpen}
          showArrow
          close={(e) => this.setState({ isOpen: false })}
          anchorNode={this.anchor}
        >
          {popoverContent}
        </EventOverlay>
        <br></br>
        <br></br>
        <br></br>
        <br></br>
      </div>
    );
  }
}
 **/
