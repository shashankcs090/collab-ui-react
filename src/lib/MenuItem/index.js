import React from 'react';
import PropTypes from 'prop-types';
import { EventOverlay, ListItem, Icon } from '@collab-ui/react/';

export default class MenuItem extends React.Component {
  static displayName = 'MenuItem';

  state = {
    anchorRef: null,
  };

  handleKeyDown = e => {
    const { handleKeyDown } = this.context;
    const { index } = this.props;
    handleKeyDown && handleKeyDown(index, this, e);
  };
  
  handleClick = e => {
    const { handleClick } = this.context;
    const { index } = this.props;
    handleClick && handleClick(index, this, e);
  };
  
  render () {
    const {
      isOpen,
      focus,
      isHeader,
      disabled,
      value,
      className,
      ariaLabel,
      subMenu,
      selectedValue,
      separator,
    } = this.props;
    return (
      <div
        className={
          'cui-menu-item' +
          `${(className && ` ${className}`) || ''}`
        }
        aria-expanded={isOpen}
        aria-haspopup={!!subMenu}
      >
        <ListItem
          onClick={this.handleClick}
          onKeyDown={this.handleKeyDown}
          ref={ref => !this.state.anchorRef && this.setState({anchorRef: ref})}
          focus={focus}
          disabled={disabled}
          isReadOnly={isHeader}
          active={isOpen}
          value={value}
          role="menuitem"
          focusOnLoad
          separator={separator}
        >
          <div className="cui-menu-item__content">
            {this.props.children}
          </div>
          <div className="cui-menu-item__selected-value" title={selectedValue}>
            {subMenu && selectedValue}
          </div>
          <div className="cui-menu-item__arrow">
            {subMenu && <Icon name="arrow-right_16"/>}
          </div>
        </ListItem>
        { isOpen ?
        <EventOverlay
          anchorNode={this.state.anchorRef}
          isOpen={isOpen}
          direction="right-top"
          closeOnClick={false}
        >
          <div
            aria-label={ariaLabel}
            role="menu"
          >
            {subMenu}
          </div>
        </EventOverlay> : null }
      </div>
    );
  }
}

MenuItem.propTypes = {
  children: PropTypes.node,
  subMenu: PropTypes.array,
  isOpen: PropTypes.bool,
  focus: PropTypes.bool,
  isHeader: PropTypes.bool,
  disabled: PropTypes.bool,
  value: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.object,
    PropTypes.array
  ]),
  selectedValue: PropTypes.string,
  className: PropTypes.string,
  shouldCloseMenu: PropTypes.bool,
  ariaLabel: PropTypes.string,
  index: PropTypes.array,
  menuItem: PropTypes.array,
  separator: PropTypes.bool,
};

MenuItem.contextTypes = {
  handleClick: PropTypes.func,
  handleKeyDown: PropTypes.func,
};

MenuItem.defaultProps = {
  children: null,
  subMenu: null,
  isOpen: false,
  focus: false,
  isHeader: false,
  disabled: false,
  value: '',
  className: '',
  shouldCloseMenu: false,
  ariaLabel: '',
  selectedValue: '',
  separator: false,
};
