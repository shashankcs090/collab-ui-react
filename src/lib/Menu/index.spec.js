import React from 'react';
import { shallow, mount } from 'enzyme';
import Menu from '../Menu';
import MenuItem from '../MenuItem';

describe('tests for <Menu />', () => {
  it('should render a Menu', () => {
    const wrapper = shallow(
      <Menu>
        <MenuItem>one</MenuItem>
      </Menu>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should focus first non disabled/ non readOnly menuItem', () => {
    const wrapper = shallow(
      <Menu>
        <MenuItem isHeader>one</MenuItem>
        <MenuItem disabled>two</MenuItem>
        <MenuItem>three</MenuItem>
      </Menu>
    );
    expect(wrapper.state().menuIndex).toEqual([2]);
    expect(wrapper.props().children[2].props.focus).toEqual(true);
  });

  it('should open/select the submenu', () => {
    const subMenu = [<MenuItem key="0">one-one</MenuItem>];
    let selectedIndex;
    const onSelect = (e, i) => selectedIndex = i;
    const wrapper = mount(
      <Menu onSelect={onSelect}>
        <MenuItem subMenu={subMenu}>one</MenuItem>
        <MenuItem>two</MenuItem>
        <MenuItem>three</MenuItem>
      </Menu>
    );
    let menuItem = wrapper.find('MenuItem').at(0);
    // click first menu item
    menuItem.find('.cui-list-item').simulate('click');
    expect(selectedIndex).toEqual([0]);
    expect(wrapper.state().menuIndex).toEqual([0, 0]);

    // click on first subMenu Item
    menuItem = wrapper.find('.cui-menu-item').at(0);
    menuItem.find('.cui-event-overlay__children .cui-list-item').simulate('click');
    expect(selectedIndex).toEqual([0, 0]);
    expect(wrapper.state().menuIndex).toEqual([0]);
    expect(wrapper.find('MenuItem').at(0).props().focus).toEqual(true);
    expect(wrapper.find('MenuItem').at(0).props().isOpen).toEqual(false);
  });

  it('should handle key-board keys', () => {
    let selectedIndex;
    const onSelect = (e, i) => selectedIndex = i;
    const subMenu = [<MenuItem key="0">one-one</MenuItem>];
    const wrapper = mount(
      <Menu onSelect={onSelect}>
        <MenuItem subMenu={subMenu}>one</MenuItem>
        <MenuItem>two</MenuItem>
        <MenuItem>three</MenuItem>
      </Menu>
    );

    // press down arrow
    let item = wrapper.find('MenuItem').at(0).find('.cui-list-item');
    item.simulate('keyDown', { which: 40 });
    expect(wrapper.state().menuIndex).toEqual([1]);

    // press up arrow
    item = wrapper.find('MenuItem').at(1).find('.cui-list-item');
    item.simulate('keyDown', { which: 38 });
    expect(wrapper.state().menuIndex).toEqual([0]);

    // press right arrow
    item = wrapper.find('MenuItem').at(0).find('.cui-list-item');
    item.simulate('keyDown', { which: 39 });
    expect(selectedIndex).toEqual([0]);
    expect(wrapper.state().menuIndex).toEqual([0, 0]);
    const subMenuItem = wrapper.find('MenuItem').at(0).find('.cui-event-overlay__children .cui-list-item');

    // press left arrow
    subMenuItem.simulate('keyDown', { which: 37 });
    expect(wrapper.state().menuIndex).toEqual([0]);
  });

  it('when children are not MenuItem should throw error', () => {
    try {
      mount(
        <Menu>
          <div>one</div>
        </Menu>
      );
    } catch(e) {
      expect(e.message).toEqual('child elements and subMenu items should be MenuItem');
    }
    try {
      mount(
        <Menu>
          <MenuItem subMenu={[<div key="0">one-one</div>]}>one</MenuItem>
        </Menu>
      );
    } catch(e) {
      expect(e.message).toEqual('child elements and subMenu items should be MenuItem');
    }
  });

});
