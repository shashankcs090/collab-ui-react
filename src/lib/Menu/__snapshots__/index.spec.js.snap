// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for <Menu /> should render a Menu 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Menu
    ariaLabel=""
    className=""
    onSelect={null}
  >
    <MenuItem
      ariaLabel=""
      className=""
      disabled={false}
      focus={false}
      isHeader={false}
      isOpen={false}
      selectedValue=""
      separator={false}
      shouldCloseMenu={false}
      subMenu={null}
      value=""
    >
      one
    </MenuItem>
  </Menu>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "aria-label": "",
      "children": Array [
        <MenuItem
          ariaLabel=""
          className=""
          disabled={false}
          focus={true}
          index={
            Array [
              0,
            ]
          }
          isHeader={false}
          isOpen={false}
          selectedValue=""
          separator={false}
          shouldCloseMenu={false}
          subMenu={null}
          value=""
        >
          one
        </MenuItem>,
      ],
      "className": "cui-menu",
      "role": "menubar",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": ".0",
        "nodeType": "class",
        "props": Object {
          "ariaLabel": "",
          "children": "one",
          "className": "",
          "disabled": false,
          "focus": true,
          "index": Array [
            0,
          ],
          "isHeader": false,
          "isOpen": false,
          "selectedValue": "",
          "separator": false,
          "shouldCloseMenu": false,
          "subMenu": null,
          "value": "",
        },
        "ref": null,
        "rendered": "one",
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "aria-label": "",
        "children": Array [
          <MenuItem
            ariaLabel=""
            className=""
            disabled={false}
            focus={true}
            index={
              Array [
                0,
              ]
            }
            isHeader={false}
            isOpen={false}
            selectedValue=""
            separator={false}
            shouldCloseMenu={false}
            subMenu={null}
            value=""
          >
            one
          </MenuItem>,
        ],
        "className": "cui-menu",
        "role": "menubar",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": ".0",
          "nodeType": "class",
          "props": Object {
            "ariaLabel": "",
            "children": "one",
            "className": "",
            "disabled": false,
            "focus": true,
            "index": Array [
              0,
            ],
            "isHeader": false,
            "isOpen": false,
            "selectedValue": "",
            "separator": false,
            "shouldCloseMenu": false,
            "subMenu": null,
            "value": "",
          },
          "ref": null,
          "rendered": "one",
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
